Exiled.Events/Patches/Events/Scp049/StartingRecall.cs

```diff
@@ -1,79 +1,80 @@
+// Licensed under the CC BY-SA 3.0 license.
+// </copyright>
+// -----------------------------------------------------------------------
+
+namespace Exiled.Events.Patches.Events.Scp049
+{
+    using System.Collections.Generic;
+    using System.Reflection.Emit;
+
+    using API.Features;
+    using API.Features.Pools;
+    using Exiled.Events.Attributes;
+    using Exiled.Events.EventArgs.Scp049;
+
+    using HarmonyLib;
+
+    using PlayerRoles.PlayableScps.Scp049;
+
+    using static HarmonyLib.AccessTools;
+
+    /// <summary>
+    ///     Patches <see cref="Scp049ResurrectAbility.ServerValidateBegin" />.
+    ///     Adds the <see cref="Handlers.Scp049.StartingRecall" /> event.
+    /// </summary>
+    [EventPatch(typeof(Handlers.Scp049), nameof(Handlers.Scp049.StartingRecall))]
+    [HarmonyPatch(typeof(Scp049ResurrectAbility), nameof(Scp049ResurrectAbility.ServerValidateBegin))]
+    internal static class Scp049ResurrectAbilityPatch
+    {
+        private static IEnumerable<CodeInstruction> Transpiler(IEnumerable<CodeInstruction> instructions, ILGenerator generator)
+        {
+            List<CodeInstruction> newInstructions = ListPool<CodeInstruction>.Pool.Get(instructions);
+
+            Label continueLabel = generator.DefineLabel();
+            Label retLabel = generator.DefineLabel();
+
+            newInstructions.InsertRange(0, new CodeInstruction[]
+            {
+                // Player player = Player.Get(this.Owner);
+                new(OpCodes.Ldarg_0),
+                new(OpCodes.Callvirt, PropertyGetter(typeof(Scp049ResurrectAbility), nameof(Scp049ResurrectAbility.Owner))),
+                new(OpCodes.Call, Method(typeof(Exiled.API.Features.Player), nameof(Exiled.API.Features.Player.Get), new[] { typeof(ReferenceHub) })),
+
+                // Ragdoll doll = Ragdoll.Get(ragdoll);
+                new(OpCodes.Ldarg_1),
+                new(OpCodes.Call, Method(typeof(Ragdoll), nameof(Ragdoll.Get), new[] { typeof(BasicRagdoll) })),
+
+                // true
+                new(OpCodes.Ldc_I4_1),
+
+                // StartingRecallEventArgs ev = new StartingRecallEventArgs(player, doll, true);
+                new(OpCodes.Newobj, GetDeclaredConstructors(typeof(StartingRecallEventArgs))[0]),
+                new(OpCodes.Dup),
+
+                // Handlers.Scp049.OnStartingRecall(ev);
+                new(OpCodes.Call, Method(typeof(Handlers.Scp049), nameof(Handlers.Scp049.OnStartingRecall))),
+
+                // if (!ev.IsAllowed)
+                //      return 1;
+                // else
+                //      continue;
+                new(OpCodes.Callvirt, PropertyGetter(typeof(StartingRecallEventArgs), nameof(StartingRecallEventArgs.IsAllowed))),
+                new(OpCodes.Brtrue_S, continueLabel),
+                new(OpCodes.Ldc_I4_1),
+                new CodeInstruction(OpCodes.Ret).WithLabels(retLabel),
+                
+                new CodeInstruction(OpCodes.Nop).WithLabels(continueLabel),

```

Comment:
Same as above

Reply:
Because this is patched at ```ServerValidateBegin```. This method need a error code that will end the whole event.
```csharp
	public override void ServerProcessCmd(NetworkReader reader)
	{
		base.ServerProcessCmd(reader);
		Vector3 position = base.ScpRole.FpcModule.Position;
		_syncRagdoll = reader.ReadNetworkBehaviour<DynamicRagdoll>();
		if (_syncRagdoll == null)
		{
			if (IsInProgress)
			{
				_errorCode = ServerValidateCancel();
				if (_errorCode != 0)
				{
					ServerSendRpc(toAll: true);
				}
				else
				{
					IsInProgress = false;
				}
			}
		}
		else
		{
			if (IsInProgress || !IsCorpseNearby(position, _syncRagdoll, out var ragdollPosition))
			{
				return;
			}
			Transform ragdollTransform = _ragdollTransform;
			BasicRagdoll curRagdoll = CurRagdoll;
			_ragdollTransform = ragdollPosition;
			CurRagdoll = _syncRagdoll;
			_errorCode = ServerValidateBegin(_syncRagdoll);
			bool flag = _errorCode != 0;
			if (flag || !ServerValidateAny())
			{
				_ragdollTransform = ragdollTransform;
				CurRagdoll = curRagdoll;
				if (flag)
				{
					ServerSendRpc(toAll: true);
				}
			}
			else
			{
				IsInProgress = true;
			}
		}
	}
```
As you can see , ```			_errorCode = ServerValidateBegin(_syncRagdoll);```.
The continueLabel is used to continue the method while the ```IsAllowed``` is True.
The retLabel is used to the return 1 value to makes error and end the event method, or the method ```ServerProcessCmd```.

Reply:
```
IL_0000: Local var 0: PlayerRoles.PlayableScps.Scp049.Scp049ResurrectAbility/ResurrectError
IL_0000: // start original
IL_0000: ldarg.0
IL_0001: callvirt   ReferenceHub PlayerRoles.PlayableScps.Subroutines.ScpStandardSubroutine`1<PlayerRoles.PlayableScps.Scp049.Scp049Role>::get_Owner()
IL_0006: call       static Exiled.API.Features.Player Exiled.API.Features.Player::Get(ReferenceHub referenceHub)
IL_000B: ldarg.1
IL_000C: call       static Exiled.API.Features.Ragdoll Exiled.API.Features.Ragdoll::Get(PlayerRoles.Ragdolls.BasicRagdoll ragdoll)
IL_0011: ldc.i4.1
IL_0012: newobj     System.Void LimitZombie.Events.EventArgs.Scp049.StartingRecallEventArgsMod::.ctor(Exiled.API.Features.Player player, Exiled.API.Features.Ragdoll ragdoll, System.Boolean isAllowed)
IL_0017: dup
IL_0018: call       static System.Void Exiled.Events.Handlers.Scp049::OnStartingRecall(Exiled.Events.EventArgs.Scp049.StartingRecallEventArgs ev)
IL_001D: callvirt   virtual System.Boolean LimitZombie.Events.EventArgs.Scp049.StartingRecallEventArgsMod::get_IsAllowed()
IL_0022: brtrue =>  Label2
IL_0027: ldc.i4.1
IL_0028: Label3
IL_0028: br =>      Label4
IL_002D: Label2
IL_002D: nop
IL_002E: ldarg.0
IL_002F: ldarg.1
IL_0030: call       PlayerRoles.PlayableScps.Scp049.ResurrectError PlayerRoles.PlayableScps.Scp049.Scp049ResurrectAbility::CheckBeginConditions(PlayerRoles.Ragdolls.BasicRagdoll ragdoll)
IL_0035: stloc.0
IL_0036: ldloc.0
IL_0037: brfalse => Label0
IL_003C: ldloc.0
IL_003D: conv.u1
IL_003E: br =>      Label5
IL_0043: Label0
IL_0043: ldarg.0
IL_0044: callvirt   virtual System.Boolean PlayerRoles.PlayableScps.Scp049.RagdollAbilityBase`1<PlayerRoles.PlayableScps.Scp049.Scp049Role>::ServerValidateAny()
IL_0049: brtrue =>  Label1
IL_004E: ldc.i4.1
IL_004F: br =>      Label6
IL_0054: Label1
IL_0054: ldc.i4.0
IL_0055: // end original
IL_0055: Label3
IL_0055: Label4
IL_0055: Label5
IL_0055: Label6
IL_0055: ret
DONE
```
Final IL code , you can check the logic here.

Reply:
I just can't find where you assign this label to any CodeInstruction. Code can be continued without this label

Reply:
```            Label continueLabel = generator.DefineLabel();```

Reply:
> I just can't find where you assign this label to any CodeInstruction. Code can be continued without this label

idk , maybe , i didnt test that.

Reply:
> ```            Label continueLabel = generator.DefineLabel();```

Yeah, you defined it but it's not attached to anything. Have you tested your transpiler?

Reply:
> > `            Label continueLabel = generator.DefineLabel();`
> 
> Yeah, you defined it but it's not attached to anything. Have you tested your transpiler?

sure. I uploaded since i fixed and tested.
And didnt i put the patched IL code ? thats the evidence lol.
It's working as what it should be rn.